can_tools/scrapers/official/base.py

```diff
@@ -116,6 +116,116 @@ def _put_exec(self, engine: Engine, df: pd.DataFrame) -> None:
 
         return worked, rows_inserted, rows_deleted
 
+    def _reshape_variables(
+        self, data: pd.DataFrame, variable_map: Dict[str, CMU]
+    ) -> pd.DataFrame:
+        """Reshape columns in data to be long form definitions defined in `variable_map`.
+
+        Parameters
+        ----------
+        data :
+            Input data
+        variable_map : Union[str,int]
+            Map from column name to output variables
+
+        Returns
+        -------
+        data:
+            Reshaped DataFrame.
+        """
+        # parse out data columns
+        value_cols = list(set(data.columns) & set(variable_map.keys()))
+        assert len(value_cols) == len(variable_map)
+        id_vars = []
+        if "location_name" in data.columns:
+            id_vars.append("location_name")
+        if "location" in data.columns:
+            id_vars.append("location")
+        if "dt" in data.columns:
+            id_vars.append("dt")
+        if "vintage" in data.columns:
+            id_vars.append("vintage")
+
+        data = (
+            data.melt(id_vars=id_vars, value_vars=value_cols)
+            .dropna()
+            .assign(
+                value=lambda x: pd.to_numeric(
+                    x["value"].astype(str).str.replace(",", "")
+                ),
+            )
+            .pipe(self.extract_CMU, cmu=variable_map)
+            .drop(["variable"], axis=1)
+        )
+
+        if "vintage" not in data.columns:
+            data["vintage"] = self._retrieve_vintage()
+
+        return data
+
+    def _rename_or_add_date_and_location(
```

Comment:
I really like this, but am thinking maybe we split implementation into two functions?

Handling locations in a func and handling dates in a func seems like good modular design to me


If we wanted to reduce number of times we call both functions we could have a simple wrapper method that does both, similar to how `fetch_normalize` does `self.normalize(self.fetch())`

Another thought is that we could have these function arguments be class attributes instead and then we could write a general `normalize` method that could be applied to most scrapers?

Reply:
In general I agree though I'm not sure it's necessary in this specific case - I'm a proponent of "functions should do one thing well" and "renaming/adding date and time columns" feels like one thing but maybe I'm overstepping there.  If it was a bit more logic I'd probably be more in favor of splitting up.

I think a more general normalize is promising - but I'm still a bit hesitant to go there without using these helper functions more and seeing what would be useful.  For example I think the tableau class `normalize` is cool - but it does look like it's pushed some more of the work into `fetch` and I'm not sure that's the right place for it theoretically.